### ISSUE TITLE
`no-base-to-string` should prevent assigning an object without `toString` to a location that expects it

### ISSUE THREAD

**Issue by bradzacher** [typescript-eslint/typescript-eslint#6051]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-base-to-string/

### Description

TS considers implicitly inherited members when computing assignability. This means that you can assign an object that does not explicitly declare `toString` to a location that expects one.

This creates a large hole in the rule that we should consider plugging by analysing assignments.

It would be a large increase in the responsibiliies of the rule which would hurt perf a bit.

### Fail

```typescript
declare const foo: {a: string};

function acceptsA(obj: {a: string, toString(): string}) {
  return `${obj}`;
}

acceptsA(foo); // error here
```

### Pass

```typescript
declare const foo: {a: string, toString(): string};

function acceptsA(obj: {a: string, toString(): string}) {
  return `${obj}`;
}

acceptsA(foo);
```

**Comment by ronami** [comment]
I like this, but I think a rule like that would be rather expensive to run. To report on the example above, in addition to possibly checking every function call, object literal fields will also need to be checked.

The closet rule that currently does something similar is `no-unsafe-argument`, and even that doesn't check object literal fields (https://github.com/typescript-eslint/typescript-eslint/issues/10158):

```ts
declare const foo: { a: any };

function acceptsA(obj: { a: string }) {
  // ...
}

acceptsA(foo); // no error
```

Should the implementation match `no-unsafe-argument` in that sense and not check object literal fields?
